### ISSUE TITLE
Taxons not showing in Edit Product

### ISSUE THREAD

**Issue by sstead** [openfoodfoundation/openfoodnetwork#1900]
Taxons aren't loading in Product Edit. I believe that by default the product category should show here, and you should be able to select a series of product categories in this field e.g. below
![image](https://user-images.githubusercontent.com/10843056/31526843-1752fcb8-b015-11e7-9a6a-8df5ad3403af.png)

![image](https://user-images.githubusercontent.com/10843056/31526792-c92bf4d6-b014-11e7-9f49-7fbbe82b7fad.png)

**Comment by KeirOsborn** [comment]
Hi Sally, 

Nice to virtually meet you. I've just started looking at this issue. 

I'd like to know a bit more about the desired behaviour here if possible. 

Is it a case of making this dropdown actually work, i.e. clicking on it displays a list of all available product categories (i.e. taxons)? If so where does this leave the product category field in the right hand column? Should this be removed?

Also it sounds like we want a way to select _multiple_ taxons, and then have them all displayed (and removable) somewhere on the page. The mechanisms to do this appear to have already been implemented in the back end to some extent - when you select a new taxon in the product categories dropdown on the right hand side of the page, and then click update, you're not actually removing the previous taxon selection but instead adding to to a list of associated taxons - the list just isn't displayed anywhere. 

Have I got this right? Or am I barking up the wrong tree?

Thanks,
Keir

**Comment by sstead** [comment]
Hi @KeirOsborn :)

This issue was just to reinstate the previous behaviour (which you've described). 
As a minimum for this issue we want the following:

a) Have all categories applied to the product show in the 'taxons' field

Currently this isn't happening. For a product that has only ever had a single category, that single category should show in the taxon field. For a product that has for example changed from the 'dairy' category to the 'specials' category, that product will have both categories, and both should show in the taxon field (this isn't happening).

b) Be able to search for and apply new categories in the taxon field

"Is it a case of making this dropdown actually work, i.e. clicking on it displays a list of all available product categories (i.e. taxons)?" YES

c) You can remove categories in the taxon field (except for the 1 selected in Category - you must have at least 1)

This was the previous behavior, which should be kept.

d) Keep the Category field as it is.

"If so where does this leave the product category field in the right hand column? Should this be removed?" no...

The two fields seem duplicate- but taxon is different because it lets users apply multiple categories- you can't do this in Category. The Category field is the one most users interact with, as it's shown in the 'new product' page, and in the BEP page (pictured below). Few users will interact with the taxons field I suspect.

![image](https://user-images.githubusercontent.com/10843056/33295434-6f38b5a2-d429-11e7-85ec-65b8bb32d508.png)

I agree it's a bit confusing for users that when they change the 'category' it does't change the category, but rather adds another in the taxon field. However, this is probably outside of the scope of this issue? As far as I know there isn't a great need to change this behaviour.  Is it a simple change to modify the Category behaviour such that 'if I change a product's category, the original category is removed and replaced with the new one'? If so I we could add this to the scope? It seems more intuitive to me, but falls into a 'nice to have but not essential' bucket.

**Comment by kirstenalarsen** [comment]
@sstead @KeirOsborn @sauloperez could this be linked to #1965? which in turn I think may be linked to Spree Upgrade, so I asked Pau to have a look at it. He is on holiday at the moment but back in action later this week. @KeirOsborn perhaps have a look and see if you think the issues are connected and if so, perhaps best wait until Pau has had a look before spending too much time on it?

**Comment by KeirOsborn** [comment]
Hi all, It looks like #1965 isn't that closely related to #1900. 

@kirstenalarsen @sauloperez Re: #1965 it seems like there's something wrong on /admin/taxonomies/# that means that you can no-longer interact with a taxonomy's associated taxons. The cryptic "Right click a child in the tree to access the menu for adding, deleting or sorting a child" alludes to how these might have been accessed before. I never saw what it looked like pre-upgrade though so I can't say for certain. 

FYI if it's really urgent you actually _can_ edit existing product categories by just hitting the URL of an individual taxon's edit action - they're apparently not linked-to anywhere but the route does exist. For example something like: /admin/taxonomies/1/taxons/3/edit. I can't find a way to create new ones though so don't know how helpful this is. 

Please correct me if i'm wrong on any of this. It's largely guesswork. 

I'm happy to work on #1965. I don't think it's particularly complicated, but it does feel like a separate issue at the moment. 

In the meantime i'll press on with #1900, treating it as unrelated unless I hear/discover otherwise. 

@sstead Re: #1900, Thanks for the info, all makes sense. 

"Is it a simple change to modify the Category behaviour such that 'if I change a product's category, the original category is removed and replaced with the new one'?" 

Yep I wouldn't have thought this would be difficult. And It definitely sounds like it would be more intuitive. If you're going to just display one of a product's taxons, it should be the primary taxon right? I'm really not that clued up on how this primary taxon is used though so do let me know if i've got this wrong. 

Thanks all,

**Comment by KeirOsborn** [comment]
@kirstenalarsen @sauloperez @sstead Scratch that. I was totally wrong. It _is_ the same issue - and indeed _is_ linked to the spree upgrade. Both are caused by a missing API key. See https://github.com/spree/spree/issues/4482. I've tested it by just setting Spree::Api::Config[:requires_authentication] = false and everything works as it should. 

The advice is to upgrade to Spree 2.2.stable but I'm guessing that isn't an option so looking for a hotfix.

**Comment by sauloperez** [comment]
Thanks a lot @KeirOsborn ! your insights provided very useful. I got the fix in https://github.com/openfoodfoundation/openfoodnetwork/pull/1984. Please, check it out.

**Comment by KeirOsborn** [comment]
Ok cool, sounds good. I was working on the basis that the desired behaviour was to use authentication when using the API, given that the requirement for authentication had been explicitly set to true. Maybe this was a mistake though. Glad we got to a solution anyway. 

For what it's worth I don't think It would be that difficult to implement a solution where authorization is used. It's used in later versions of spree so presumably we'd want to switch back to using it at a later date anyway. 

What does anyone think?

**Comment by sauloperez** [comment]
It was us that explicitly set it to true in the last minor upgrade of the Spree that we did, and we might need to remember the actual reason, but the Spree version we are at now has a bug that doesn't allow using API auth. We'll have to get to 2.2.1 (if I'm not mistaken), to re-enable it back.

**Comment by KeirOsborn** [comment]
Hi @sstead, I'm guessing the extra changes we discussed here are no-longer a priority. I.e. this: 

"Is it a simple change to modify the Category behaviour such that 'if I change a product's category, the original category is removed and replaced with the new one'?" 

This remains true but to be honest I suggested the changes before I understood what the solution to issue #1900 would actually look like. Now that I understand what's going on I agree that this is outside of the scope of this issue and so i'm not going to work on it for the time being if that's ok with everyone. Let me know if you feel differently though. 

Thanks,
Keir
